scapy/contrib/isotp.py

```diff
@@ -57,100 +52,1171 @@ def answers(self, other):
+            self.ready = []
+            self.ctors = {}
+
+        def feed_first_frame(self, identifier, ea, data):
+            if len(data) < 3:
+                # At least 3 bytes are necessary: 2 for length and 1 for data
+                return False
+
+            ex_l = struct.unpack('>H', bytes(data[:2]))[0]
+            ex_l = ex_l & 0x0fff
+            isotp_data = data[2:]
+
+            key = (ea, identifier, 1)
+            self.ctors[key] = self.Ctor(ex_l, isotp_data)
+            return True
+
+        def feed_single_frame(self, identifier, ea, data):
+            if len(data) < 2:
+                # At least 2 bytes are necessary: 1 for length and 1 for data
+                return False
+
+            length = six.indexbytes(data, 0) & 0x0f
+            isotp_data = data[1:length + 1]
+
+            if length > len(isotp_data):
+                # CAN frame has less data than expected
+                return False
+
+            self.ready.append((identifier, ea, isotp_data))
+            return True
+
+        def feed_consecutive_frame(self, identifier, ea, data):
+            if len(data) < 2:
+                # At least 2 bytes are necessary: 1 for sequence number and
+                # 1 for data
+                return False
+
+            first_byte = six.indexbytes(data, 0)
+            seq_no = first_byte & 0x0f
+            isotp_data = data[1:]
+
+            key = (ea, identifier, seq_no)
+            ctor = self.ctors.pop(key, None)
+
+            if ctor is None:
+                # There is no message constructor waiting for this frame
+                return False
+
+            ctor.push(isotp_data)
+            if ctor.ready is None:
+                # full ISOTP message is not ready yet, put it back in
+                # constructors array
+                next_seq = (seq_no + 1) % 16
+                key = (ea, identifier, next_seq)
+                self.ctors[key] = ctor
+            else:
+                self.ready.append((identifier, ea, ctor.ready))
+
+            return True
+
+        def try_feed(self, identifier, ea, data):
+            first_byte = six.indexbytes(data, 0)
+            if len(data) > 1 and first_byte & 0xf0 == 0:
+                self.feed_single_frame(identifier, ea, data)
+            if len(data) > 2 and first_byte & 0xf0 == 0x10:
+                self.feed_first_frame(identifier, ea, data)
+            if len(data) > 1 and first_byte & 0xf0 == 0x20:
+                self.feed_consecutive_frame(identifier, ea, data)
+
+        def feed(self, can):
+            identifier = can.identifier
+            data = bytes(can.data)
+
+            if len(data) > 1:
+                self.try_feed(identifier, None, data)
+            if len(data) > 2:
+                ea = six.indexbytes(data, 0)
+                self.try_feed(identifier, ea, data[1:])
+
+        def count(self):
+            return len(self.ready)
+
+        def pop(self, identifier=None, ext_addr=None):
```

Comment:
In my opinion, pop should return a ISOTP Packet

Reply:
It is an internal class, and it's not supposed to be used by the end user.
The sniff method of ISOTPSniffer will return ISOTP packets.